Provide foundation for Tests using JUnit5

c-mueller opened the pull request:
#### Short Description
- Migrates all current Tests to JUnit 5
- Adds "guidelines" how tests can be written in JUnit 5
- Adds foundation for Service tests using PostgreSQL

#### Proposed Changes
  * Migrates all exisiting Controller tests to JUnit 5
  * WebMvc Controller tests now use `@WebMvcTest` instead of `@SpringBootTest`
  * Moves JPA configuration in a separate class to ensure Tests can be loaded without Jpa.
  * Updates Travis to use Ubuntu 18.04 instead of 14.04
  * Adds classes to allow easy implementation of service tests using PostgreSQL
  * Adds basic docs on how to write tests, both for services and controllers 

**Fixes**: #PlanQK/q-tal#30

timniederhausen commented:
Regarding the `@JsonAnyGetter` and `@JsonAnySetter` stuff: Do we need to deserialize the JSON again? Comparing it with the "expected" JSON response should be enough.

Given that we're going to switch away from the `*ListDto` classes this would make changes easier in the future.
Also, can we even deserialize the upcoming PagedModels?

Depending on how much work that is, this might be something for the next sprint.

c-mueller commented:
> Regarding the `@JsonAnyGetter` and `@JsonAnySetter` stuff: Do we need to deserialize the JSON again?

I actually found a fix for this, the throwing of such exceptions can be prevented by setting a Flag in the ObjectMappers configuration. Doing so will make the `@JsonAnyGetter/Setter` annotations uneccessary. For testing purposes this is perfect. The only things we cannot validate using this approach are the links and the resources in `_embedded` for these we have to find special workarounds.

Validating links for example can be done using the `JSONObject` class, as shown in the root controller test.

> Comparing it with the "expected" JSON response should be enough.

I don't think this approach is suitable, since it is not very error resistent, since order of the objects becomes relevant, if the payload is identical but in a different order, comparing against the expected will fail. A similar problem might occur if we have to rename a field...


> Given that we're going to switch away from the `*ListDto` classes this would make changes easier in the future.
> Also, can we even deserialize the upcoming PagedModels?

We can partially deserialize these using some "workarounds". Assuming the response looks similar to this:
![grafik](https://user-images.githubusercontent.com/2706976/83432152-be891480-a438-11ea-9d46-0b83d59c9201.png)


To deserialize/validate this we have several options:
1. Write a custom DTO only for validating in tests
2. Use JsonPath to get the elements needed for validation
    - More Info: https://github.com/json-path/JsonPath
3. Use JSONObjects and JSONArrays to get to the desired parts of the test

Approches No. 2 and 3 are probably the simpler ones here.

timniederhausen commented:
> I don't think this approach is suitable, since it is not very error resistent, since order of the objects becomes relevant, if the payload is identical but in a different order, comparing against the expected will fail. A similar problem might occur if we have to rename a field...

Good point, changing field names could force us to change unrelated tests as well.
The only point where object ordering matters though is in arrays / lists, for which we (should) have consistent ordering anyway.

Another take on the idea would be to have an "expected" `PagedModel<T>` / `EntityModel<T>` / ..., serialize that in the test and compare it to the controller's response. This way we don't have to hard-code their JSON representation. There are quite a few libraries that could do this for us (for example, [JsonUnit](https://github.com/lukas-krecan/JsonUnit))

c-mueller commented:
> Another take on the idea would be to have an "expected" `PagedModel<T>` / `EntityModel<T>` / ..., serialize that in the test and compare it to the controller's response. This way we don't have to hard-code their JSON representation. There are quite a few libraries that could do this for us (for example, [JsonUnit](https://github.com/lukas-krecan/JsonUnit))

I think this sounds like the most suitable solution, assuming the `ObjectMapper` serializes these objects properly. I can take a look at that tomorrow.

c-mueller commented:
> > Another take on the idea would be to have an "expected" `PagedModel<T>` / `EntityModel<T>` / ..., serialize that in the test and compare it to the controller's response. This way we don't have to hard-code their JSON representation. There are quite a few libraries that could do this for us (for example, [JsonUnit](https://github.com/lukas-krecan/JsonUnit))
> 
> I think this sounds like the most suitable solution, assuming the `ObjectMapper` serializes these objects properly. I can take a look at that tomorrow.

Update: this does not seem to work, I've added a Util method to handle the unmarshalling of paged models using JSONObjects. When testing you simply hand over the response as string, the key under witch the resources can be expected, an instance of a ObjectMapper and the class used for Unmarshalling. As a result a list, in the identical order will be returned. The only missing thing in the list are the links, these will have to be verified manually as discussed above.

I've also added a utility method to retrieve the page information from the paged model response.

T-Haller commented on docs/dev/writing-tests.md:
```diff
@@ -0,0 +1,159 @@
+
+#### Getting the objects from a `PagedModel`
+
+If you want to verify the contents of a PagedModel you can use a small helper method in the `ObjectMapperUtils` class: `ObjectMapperUtils.mapResponseToList()`. To retrieve a list of DTOs, just pass the response as a string, returned from MockMvc, as well as the key under which the resources are located. The key in the case, the class name ends with DTO usually is `<ClassName>es` where the class n ...
```
Here is a word too much: `you must pass the an Instance of`

c-mueller replied:
Fixed!
